### ISSUE TITLE
[Bug] tl.dot(..., allow_tf32=False) returns NaN/Inf in recompute_w_u_fwd_kernel_npu with fully-finite bf16 inputs

### ISSUE THREAD

**Issue by TimeVShow** [fla-org/flash-linear-attention#1215]
### Checklist

- [x] I have checked [FAQs](https://github.com/fla-org/flash-linear-attention/blob/main/FAQs.md) and existing issues for similar problems
- [x] Please report this bug in English to ensure wider understanding and support

### Describe the Bug

On Ascend 910A, `tl.dot(..., allow_tf32=False)` inside the WY-representation recomputation of
`chunk_gated_delta_rule` produces **NaN/Inf** even though **both operands are fully finite** (the bf16
inputs are cast to fp32 before the dot).

The defect is localized to two `tl.dot` calls in
[`fla/ops/gated_delta_rule/backends/triton_ascend/wy_fast.py`](https://github.com/fla-org/flash-linear-attention/blob/main/fla/ops/gated_delta_rule/backends/triton_ascend/wy_fast.py),
inside `recompute_w_u_fwd_kernel_npu`:

```python
b_u = tl.dot(b_A, b_vb, allow_tf32=False)   # (64x64) @ (64x64)
b_w = tl.dot(b_A, b_kb, allow_tf32=False)   # (64x64) @ (64x64)
```

Two independent triggers both have **inputs with `nan = 0` and `inf = 0`**, yet the `tl.dot` output
contains NaN/Inf. Recomputing the identical matmul on CPU in fp32 is fully finite. Bad values cluster
**only on heads 0 / 16 / 24** (fixed NPU cores/tiles), which is inconsistent with a data-driven overflow
or a NaN source and points to a hardware/compiler-level `tl.dot` correctness issue on this tile shape.

### Kernel constants / tile geometry

For the two failing `tl.dot` calls, the relevant kernel constants are:

| Constant | Value |
|---|---|
| `BT` (tile along time) | 64 |
| `BK` (tile along key dim) | 64 |
| `BV` (tile along value dim) | 64 |
| `K` | 128 |
| `V` | 128 |
| `H` / `HV` | 32 |

So both dots are a `(64, 64) @ (64, 64)` fp32 matmul (`allow_tf32=False`) whose two operands are fp32 (the bf16 inputs are cast to fp32 right before the dot).

### The two dot operands

A confirmed bad output element is `u[0, 1, 24, 78]` (block-local `[1, 14]`). Its operands are:

| Operand | Definition | Shape / dtype | Value range |
|---|---|---|---|
| `b_A` (lhs, shared by both dots) | the bf16 `solve_tril` output `A[0, 0:64, 24, :]`, cast to fp32 (`tl.load(...).to(tl.float32)`) | `(64, 64)` fp32 | **unit lower-triangular**: diagonal = 1, strict upper = 0; lower entries are ≈ 49% zero, otherwise roughly Gaussian with `std ≈ 2.43`, total range `[-24.75, +22.0]` |
| `b_vb` (rhs of the `u` dot) | `b_v * b_beta[:, None]`, where `b_v = fp32(v[0, 0:64, 24, 64:128])` and `b_beta = fp32(beta[0, 0:64, 24])` | `(64, 64)` fp32 | small values: range `[-0.244, +0.471]`, `mean ≈ 0.004`, `std ≈ 0.05` |
| `b_kb` (rhs of the `w` dot) | `b_k * b_beta[:, None]` (plus `* exp2(g)[:, None]` when gating is enabled), `b_k = fp32(k)` block | `(64, 64)` fp32 | same small-`O(0.05)` scale, fully finite |

The key observation is that both operands are **small, well-conditioned, fully finite fp32 values**
(with bf16 provenance): `amax(|b_A|) ≈ 25`, `amax(|b_vb|) ≈ 0.5`, so the exact fp32 result is at most
`O(64 * 25 * 1) ≈ 1e3` per element. No legitimate intermediate can approach the fp32 overflow limit.

### How to generate an equivalent trigger

Because the operands are ordinary bf16 data (a unit-lower-triangular matrix with `std ≈ 2.4`, and a
small-scale `O(0.05)` right-hand side), any finite data in the same range reproduces the fault. A
minimal standalone recipe:

```python
import torch

BT = BK = BV = 64
torch.manual_seed(0)

# b_A: (64,64) bf16 unit-lower-triangular, std ~2.4, range ~[-25, 25]
L = torch.randn(BT, BT, dtype=torch.float32) * 2.43
L = torch.tril(L, -1).clamp(-25.0, 25.0)
b_A = (L + torch.eye(BT, dtype=torch.float32)).to(torch.bfloat16)

# b_vb: (64,64) bf16, small-scale rhs, range ~[-0.25, 0.5]
b_vb = (torch.randn(BT, BV, dtype=torch.float32) * 0.05).clamp(-0.25, 0.5).to(torch.bfloat16)

# CPU reference is always finite
u_cpu = b_A.float() @ b_vb.float()
assert torch.isfinite(u_cpu).all()
```

Running `tl.dot(b_A, b_vb, allow_tf32=False)` on Ascend for block `[0, 1, 24]` — i.e. within
`recompute_w_u_fwd_kernel_npu` — instead yields non-finite values on heads 0/16/24, while the
same operands recomputed in fp32 (`allow_tf32=False`) on CPU are finite.

### Inputs / outputs

All dump tensors are fully finite except the `tl.dot` output `u`:

| Tensor | N105 shape | N105 nan/inf | N71 shape | N71 nan/inf |
|---|---|---|---|---|
| `A_kkt` (fp32) | (1,120,32,64) | 0 / 0 | (1,108,32,64) | 0 / 0 |
| `A_solve` (bf16) | (1,120,32,64) | 0 / 0 | (1,108,32,64) | 0 / 0 |
| `A` (bf16, dot lhs) | (1,120,32,64) | 0 / 0 | (1,108,32,64) | 0 / 0 |
| `v` (bf16) | (1,120,32,128) | 0 / 0 | (1,108,32,128) | 0 / 0 |
| `beta` (bf16) | (1,120,32) | 0 / 0 | (1,108,32) | 0 / 0 |
| `k` (bf16) | (1,120,32,128) | 0 / 0 | (1,108,32,128) | 0 / 0 |
| `g` (fp32) | (1,120,32) | 0 / 0 | (1,108,32) | 0 / 0 |
| **`u` (dot output)** | (1,120,32,128) | **50 / 1** | (1,108,32,128) | **49 / 1** |

Value ranges of the inputs (finite, well-conditioned):

| Tensor | N105 value range | N71 value range |
|---|---|---|
| `A` (bf16) | `amax ≈ 10.19` | `amax ≈ 1.0` |
| `v` (bf16) | small, finite | small, finite |
| `beta` (bf16) | small, finite | small, finite |
| `k` (bf16) | small, finite | small, finite |

The bad blocks in `A` are unit-lower-triangular (diag=1), consistent with the `solve_tril` step.

### Bad values cluster on fixed heads (decisive)

For `B=1`, `i_bh = task_id % (B *HV)` and `task_id % 32` is the head; each `task_id` maps to a
specific NPU core.

| Head | N105 bad count | N71 bad count |
|---|---|---|
| **0** | 16 | 15 |
| **16** | 15 | 16 |
| **24** | 20 | 19 |
| other 29 heads | **0** | **0** |

Bad outputs appear only on heads 0/16/24 — the same three fixed cores across both runs — while the
other 29 cores produce fully correct `tl.dot` results. A NaN source or arithmetic overflow would spread across all heads and contiguous rows/columns, so this fixed-core pattern is the fingerprint of a hardware/compiler-level `tl.dot` defect, not a data problem.

The same head-0 tile fails in both runs despite very different input magnitudes (`amax(A)=10.19` on N105 vs `amax(A)=1.0` on N71), confirming failure correlates with the tile/core index, not with operand magnitude.

### CPU reference is fully finite

```
u_REF (cpu fp32)  N105: 0 / 491520 non-finite, range [-26.6242, 19.3365]
u_REF (cpu fp32)  N71:  0 / 442368 non-finite, range [-3.9237,  6.54224]
```

The theoretical accumulation bound for `BT=64` is ≈ 2.97e4 (N105) / ≈ 2.7e4 (N71), ~34 orders of magnitude below `fp32` max (3.4e38). The NPU output additionally contains large corrupted finite values on the order of `±3.4e38`, which cannot be legitimate intermediate magnitudes.

### Sample bad coordinates (actual vs CPU reference)

N105:

| (b,t,h,v) | actual | cpu ref |
|---|---|---|
| (0,1,24,78) | nan | -0.024849891662597656 |
| (0,2,0,70) | nan | -0.0195681 |
| (0,5,24,12) | nan | 0.0591132 |
| (0,50,24,112) | nan | -0.365311 |
| (0,53,24,60) | nan | -3.52666 |
| (0,112,16,58) | nan | -3.70299 |
| (0,117,16,54) | **inf** | -0.0749135 |

N71:

| (b,t,h,v) | actual | cpu ref |
|---|---|---|
| (0,2,0,70) | nan | -0.0199423 |
| (0,12,24,102) | nan | 0.0219586 |
| (0,84,16,36) | **inf** | 0.0538126 |
| (0,107,16,70) | nan | 0.0225604 |

## Reproduction data summary

No dumps are attached. To reproduce, generate finite bf16 data matching the ranges above:

- `A`: `(1, T, 32, 64)` bf16, each `( t, h)` block unit-lower-triangular, entries `std ≈2.4`, range `[-25,25]`;
- `v`: `(1, T, 32, 128)` bf16, small finite values;
- `beta`: `(1, T, 32)` bf16, small finite values;
- `k`: `(1, T, 32, 128)` bf16, small finite values;
- `g`: `(1, T, 32)` fp32, finite.

with `T =120` (or `108`) and `B =1`. Then run `chunk_gated_delta_rule` forward and inspect `u` on heads 0/16/24; those tiles contain the NaN/Inf while all inputs remain finite.

### Expected Behavior

infinite value

## Environment

| Component | Version |
|---|---|
| Hardware | Ascend 910B (`Ascend910_9382`), 4 nodes, 1 card per container |
| Driver | `25.3.rc1.2` |
| CANN | `9.1.0` (`/usr/local/Ascend/cann -> cann-9.1.0`) |
| Python | `3.12.13` |
| torch | `2.10.0+cpu` |
| torch_npu | `2.10.0.post2` |
| triton (imported module) | `3.2.0` |
| triton-ascend (dist) | `3.2.2` |
| transformers | `5.15.1` |
| ms-swift | `4.5.2` |
| megatron-core | `0.19.0` |
| flash-linear-attention (FLA) | `0.6.0` @ `5e02dd3a7651f5f2797eb8b12bbec401826031e1` |

Workload: Qwen3.5-397B distributed full-SFT (KDA path), epoch 0.

**Comment by zheliuyu** [comment]
> On Ascend 910A, tl.dot(..., allow_tf32=False) inside the WY-representation recomputation of
chunk_gated_delta_rule produces NaN/Inf even though both operands are fully finite (the bf16
inputs are cast to fp32 before the dot).

Please use `npu-smi info` to verify whether the hardware model is the 910A.

**Comment by sunyi0505** [comment]
NaN occurs because solve_tril overflows the ill‑conditioned $(I+L)^{-1}$ into Inf, which then goes into tl.dot and becomes Inf+(-Inf) / Inf*0. To resolve this issue, k needs F.normalize(..., p=2, dim=-1) or enable use_qk_l2norm_in_kernel

**Comment by zhiyuan1i** [comment]
> NaN occurs because solve_tril overflows the ill‑conditioned ( I + L ) − 1 into Inf, which then goes into tl.dot and becomes Inf+(-Inf) / Inf*0. To resolve this issue, k needs F.normalize(..., p=2, dim=-1) or enable use_qk_l2norm_in_kernel

Correct. DeltaRule needs use L2norm in K, usually both Q and K, And FLA's kernel designed with l2norm.

Could you please check your codes? and give a minimal script to reproduce the error.

**Comment by zheliuyu** [comment]
@TimeVShow Is this issue fixed? Awaiting your response.
